### PR TITLE
feat: nuxt-3 starter example

### PR DESCRIPTION
**Pull request by mdarrik**
## WHY are these changes introduced?

Example of Nuxt 3 - [ENG-4185](https://nacelle.atlassian.net/browse/ENG-4185) 

## WHAT is this pull request doing?

Adds an example of Nuxt 3 + Nacelle, using the existing Nuxt starter as a template.

- Features a composable-based state management solution instead of Vuex/Pinia.
- Shopify Cart/Checkout
- Nuxi generate support with pre-rendered routes for a static experience.
- The ReadMe also links out to relevant issues around static generation support for tracking of when generate/hybrid modes might be stable.

## How to Test

🖥 [Local development instructions](https://github.com/getnacelle/nacelle-js/blob/mdarrik/ENG-4185-Nuxt-3-starter-example/examples/nuxt-3/README.md)

1. Checkout mdarrik/ENG-4185-Nuxt-3-starter-example (or using gh CLI: `gh pr checkout 205`)
2. Run `npm run bootstrap` from root of the monorepo
3. Navigate to examples/nuxt-3/
4. Copy the `.env.example` file into a `.env` file and fill out the values (I can provide the ones I used if desired)
5. Run dev server with `npm run dev` and test pages and cart/checkout
6. Do a production static build with `npm run generate`
7. Test the build with `npx serve .output/public` and navigate to pages and test cart/checkout

### CODE REVIEW

**Review by KrisQ — CHANGES_REQUESTED** [review]
Hey @mdarrik,

Everything looks great and works when generating:

Screen Shot 2022-08-17 at 10 20 11 AM

Screen Shot 2022-08-17 at 10 20 27 AM

Screen Shot 2022-08-17 at 10 20 45 AM

However with `npm run dev` I am running into an issue with the SDK when attempting to add to cart:
Screen Shot 2022-08-17 at 10 24 22 AM

File: examples/nuxt-3/components/ProductCard.vue [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+          >{{ option.name }}</label
+        >
+        <select :id="`${option.name}-select-id`" class="product-card__select">
+          <option
+            v-for="(value, vIndex) in option.values"
+            :value="value"
+            :key="value"
+          >
+            {{ value }}
+          </option>
```
Comment (NWRichmond) [inline examples/nuxt-3/components/ProductCard.vue:38]: vue:no-unused-vars  'vIndex' is defined but never used

```suggestion
          <option v-for="value in option.values" :value="value" :key="value">
            {{ value }}
          </option>
```

File: examples/nuxt-3/composables/checkout/index.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+export async function processCheckout(): Promise<void> {
+  try {
+    const cartState = useCart();
+    const shopifyCheckoutClient = useShopifyCheckout();
+    const checkoutItems = cartState.lineItems.map((lineItem) => ({
+      quantity: lineItem.quantity,
+      variantId: lineItem.variant.id
+    }));
+    const checkoutArgs = { cartItems: checkoutItems };
```
Comment (NWRichmond) [inline examples/nuxt-3/composables/checkout/index.ts:52]: It looks like `checkoutArgs` needs a type. We could use an inline type here:
```suggestion
    const checkoutArgs: { cartItems: typeof checkoutItems; id?: string } = {
      cartItems: checkoutItems
    };
```

File: examples/nuxt-3/composables/checkout/index.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import { get, set, del } from 'idb-keyval';
+import createShopifyCheckoutClient, {
+  CheckoutClient
+} from '@nacelle/shopify-checkout';
+import { ShopifyCheckout } from '@nacelle/shopify-checkout/dist/types/checkout-client.types';
+import createCheckout from '@nacelle/shopify-checkout/dist/types/client/actions/checkoutCreate';
```
Comment (NWRichmond) [inline examples/nuxt-3/composables/checkout/index.ts:6]: Looks like `createCheckout` is an unused import.

`ShopifyCheckout` is too at the moment, but it's used in another suggestion below.

```suggestion
```

File: examples/nuxt-3/composables/checkout/index.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+export async function initCheckout(): Promise<void> {
+  try {
+    const shopifyCheckoutClient = useShopifyCheckout();
+    checkoutState.checkoutId = await get('checkoutId');
+    if (checkoutState.checkoutId) {
+      const checkout = await shopifyCheckoutClient.get({
+        id: checkoutState.checkoutId
+      });
```
Comment (NWRichmond) [inline examples/nuxt-3/composables/checkout/index.ts:33]: Property 'completed' does not exist on type 'void'

```suggestion
      const checkout = (await shopifyCheckoutClient.get({
        id: checkoutState.checkoutId
      })) as ShopifyCheckout;
```

File: examples/nuxt-3/app.vue [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  font-size: small;
+  background-color: #60a5fa;
+  clip-path: circle(35%);
+  padding: 1ch;
+  text-decoration: none !important;
+  text-decoration-style: none;
+}
+.cart-link {
+  /* display: flex; */
+  /* align-items: end; */
```
Comment (NWRichmond) [inline examples/nuxt-3/app.vue:58]:
```suggestion
```

File: examples/nuxt-3/pages/index.vue [hunk 6]
```diff
@@ -0,0 +1,44 @@
+<template>
+  <main class="home">
+    <div class="home__logo">
+      <img src="/nacelle-logo.svg" />
```
Comment (NWRichmond) [inline examples/nuxt-3/pages/index.vue:4]: Can we add some alt text to this image?

File: examples/nuxt-3/pages/products/[handle].vue [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        </button>
+      </form>
+    </article>
+  </main>
+</template>
+
+<script setup>
+import { addItemToCart } from '~~/composables/cart';
+import { ProductPageQuery } from '~~/queries';
+// import { useSdk } from '~~/composables/nacelle';
```
Comment (NWRichmond) [inline examples/nuxt-3/pages/products/[handle].vue:82]:
```suggestion
```

File: examples/nuxt-3/pages/products/[handle].vue [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  }
+};
+
+const addItem = async (e) => {
+  const cartVariant = transformProductForCart({
+    product: product,
+    variant: selectedVariant
+  });
+  if (cartVariant) {
+    //unref the quantity so only it's value gets used in the cart
```
Comment (NWRichmond) [inline examples/nuxt-3/pages/products/[handle].vue:148]:
```suggestion
    // unref the quantity so only its value gets used in the cart
```

File: examples/nuxt-3/queries/collection-page-query.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                        node {
+                            ...CollectionProductFragment
+                        }
+                    }
+                }
+            }
+        }
+    }
+    ${CollectionProductFragment}
+`;
```
Comment (NWRichmond) [inline examples/nuxt-3/queries/collection-page-query.js:51]: Might be nice on the eyes to use consistent indentation with these queries.

**Review by NWRichmond — CHANGES_REQUESTED** [review]
I'm really liking:
- TypeScript composables (the Cart composable in particular)
- Composition API usage in general
- Helpful code comments

There are some areas for improvement, most of which are small or formatting-related.

Functionality-wise, I'm having difficulty getting Add to Cart functionality working, both on PDPs and Collection PLPs:

https://user-images.githubusercontent.com/5732000/185229928-a211e189-16ce-4431-adbb-6fb84ecb007e.mov

https://user-images.githubusercontent.com/5732000/185230020-37105ce0-cecc-443d-abd8-09bb1de65828.mov

Lastly, I think we can get rid of ESLint parsing errors by adding an ESLint config file. When I copied the config from https://github.com/getnacelle/nacelle-js/blob/main/examples/nuxt-2-composition-api/.eslintrc.js, the parsing errors were resolved.

**Comment by mdarrik** [comment]
@NWRichmond @KrisQ the cart functionality should be working in local dev now. The transpilation of the sdk causes issues in the browser during local dev (but not production builds). So I set it up to only conditionally transpile during a production build. Will call that out in the README update suggested by Nick

File: examples/nuxt-3/queries/collection-page-query.js (hunk 9 above)
Comment (mdarrik) [reply examples/nuxt-3/queries/collection-page-query.js:51]: Oops. Looks like this was actually malformed gql for the second query. Fixed

File: examples/nuxt-3/pages/index.vue (hunk 6 above)
Comment (mdarrik) [reply examples/nuxt-3/pages/index.vue:4]: Done. Thanks for catching that!

File: examples/nuxt-3/README.md [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+## Features
+
+- Static prerendering of all pages
+- Global Cart/Checkout state managed with composables instead of Vuex
+- `@nacelle/shopify-checkout` to manage checking out with Shopify
+- TS for composables to provide better editor hints and DX since they're auto imported, while using JS for Vue components and other files.
+
+## Nuxt 3 Caveats
+
+- v1.x of the `@nacelle/storefront-sdk` currently exports esm as `esm.js`, which the [Nuxt 3 docs call out as an unsupported pattern](https://v3.nuxtjs.org/guide/going-further/esm#what-is-native-esm) for native Node.js ESM. The result is that the storefront sdk can be used in dev mode, but not in production builds. So to safely use the storefront sdk, it needs to be transpiled by adding it to the [`build.transpile` key of the `nuxt.config.ts file`](./nuxt.config.ts).
```
[the remark this answers is not in the document]
Comment (mdarrik) [reply examples/nuxt-3/README.md]: Attempted an update in light of the requirement that transpiration breaks dev mode.

File: examples/nuxt-3/composables/checkout/index.ts (hunk 2 above)
Comment (mdarrik) [reply examples/nuxt-3/composables/checkout/index.ts:51]: This looks like TS is able to generate an anonymous type from the constructed objects. 
VS Code typescript popup showing checkoutArgs having a type of an array of objects with a cartItems key. That cartItems key has a type of objects that contain quantity: number and variantId: string

**Comment by mdarrik** [comment]
> Lastly, I think we can get rid of ESLint parsing errors by adding an ESLint config file. When I copied the config from https://github.com/getnacelle/nacelle-js/blob/main/examples/nuxt-2-composition-api/.eslintrc.js, the parsing errors were resolved. 

So I'm not sure if the existing ESLint packages for Nuxt are quite ready for Nuxt 3 support? There are a couple open issues around Nuxt 3 that might mean it works for some ESLint errors and breaks for others.  

Eg there's [Issue 173 of `eslint-plugin-nuxt`](https://github.com/nuxt/eslint-plugin-nuxt/issues/173) and [issue 216 of `nuxt/eslint-config`](https://github.com/nuxt/eslint-config/issues/216).

**Comment by NWRichmond** [comment]
@mdarrik with respect to ESLint, I used https://github.com/nuxt/framework/discussions/2815#discussioncomment-2050408 as a starting point to see if I could get past some of the issues that you described with lining setups. While we can't safely run Nuxt 2 & Nuxt 3 linting setups in this monorepo yet (due to collisions between Vetur and Volar), the local listing setup did help to uncover a few outstanding issues, which are fixed in https://github.com/getnacelle/nacelle-js/pull/209.

I'd love to get your take on that PR and see if there are any changes in it that you think would be problematic. If you think it's good to go, then I suspect it will get `examples/nuxt-3` to a state that's indistinguishable from what it will look like when we can safely run Nuxt 2 & Nuxt 3 linting in the same monorepo.

**Review by NWRichmond — APPROVED** [review]
Nicely done, @mdarrik! Everything is working as expected now. Thank you for those updates.

https://user-images.githubusercontent.com/5732000/186260818-e4ab77fe-db07-4146-9dca-48fdc91c3e34.mov

**Pull request merged**